engine/src/test/java/io/zeebe/engine/util/ProcessExecutor.java

```diff
@@ -225,6 +229,15 @@ private void createProcessInstance(final StepStartProcessInstance startProcess)
   }
 
   private void triggerTimer(final StepTriggerTimer timerStep) {
+    final Record<TimerRecordValue> timerRecord =
+        RecordingExporter.timerRecords(TimerIntent.CREATED)
+            .withHandlerNodeId(timerStep.getTargetElementId())
+            .getFirst();
+
+    Awaitility.await(
+            "until timer " + timerStep.getTargetElementId() + " has created has been created")
+        .until(() -> engineRule.getLastWrittenPosition(1) >= timerRecord.getPosition());
```

Comment:
Here, we wait until the stream processor written the `created` event of the timer. But the timer is scheduled as a side effect that is called after the `lastWrittenPosition` is updated.
So, there is still the chance of a race condition. 

We could use the [callbacks](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/test/java/io/zeebe/engine/util/EngineRule.java#L91-L92) of the `EngineRule` (i.e. `onProcessed`/`onSkipped`) to be sure that the side effects are called before.
As suggested by Peter, we could expose the last processed position as a getter on the `EngineRule`.

Reply:
Sorry, I meant to use last processed :see_no_evil: I trusted my auto complete too much :sweat_smile: